keps/sig-cluster-lifecycle/0015-kubeadm-join-master.md

```diff
@@ -0,0 +1,431 @@
+components will be deployed on the new master node.
+As described in goals/non goals, in this first release of the proposal
+creating a new master node doesn't trigger the creation of a new etcd member on the
+same machine.
+
+#### Add a new master node (dynamic workflow)
+
+As a kubernetes administrator, (_at any time_) I want to add a new master node* to an existing
+Kubernetes cluster.
+
+### Implementation Details
+
+#### Initialize the Kubernetes cluster
+
+As of today, a Kubernetes cluster should be initialized by running `kubeadm init` on a
+first master, afterward referred as the bootstrap master.
+
+in order to support the `kubeadm join --master` workflow a new Kubernetes cluster is
+expected to satisfy following conditions :
+
+- The cluster must have a stable `controlplaneAddress` endpoint (aka the IP/DNS of the
+  external load balancer)
+- The cluster must use an external etcd.
+
+All the above conditions/settings could be set by passing a configuration file to `kubeadm init`.
+
+#### Preparing for execution of kubeadm join --master
+
+Before invoking `kubeadm join --master`, the user/higher level tools
+should copy control plane certificates from an existing master node, e.g. bootstrap master
+
+> NB. kubeadm is limited to execute actions *only*
+> in the machine where it is running, so it is not possible to copy automatically
+> certificates from remote locations.
+
+Please note that strictly speaking only ca certificate and and service account key pair
+certificate are required to be equal among all masters. Accordingly:
+
+- `kubeadm join --master` will check for the mandatory certificates and fail fast if
+  they are missing
+- If some/all of the other certificates are provided by the user as well, `kubeadm join --master`
+  will use them without further checks.
+- If any other certificates are missing, `kubeadm join --master` will create them.
+
+> see "Strategies for distributing cluster certificates" paragraph for
+> additional info about this step.
+
+#### The kubeadm join --master workflow
+
+The `kubeadm join --master` workflow will be implemented as an extension of the
+existing `kubeadm join` flow.
+
+`kubeadm join --master` will accept an additional parameter, that is the apiserver advertise
+address of the joining node; as details in following paragraphs, the value assigned to
+this parameter depends on the user choice between a dynamic bootstrap workflow or a static
+bootstrap workflow.
+
+The updated join workflow will be the following:
+
+1. Discovery cluster info [No changes to this step]
+
+   > NB This step waits for a first instance of the kube-apiserver to become ready
+   > (the bootstrap master); And thus it acts as embedded mechanism for handling the sequence
+   > `kubeadm init` and `kubeadm join` actions in case of parallel node creation.
+
+2. In case of `join --master` [New step]
+
+   1. Using the bootstrap token as identity, read the `kubeadm-config` configMap
```

Comment:
I want us to dive into whether we should use different bootstrap token roles for joining a master and a node

Reply:
Nothing prevents us to this, but we should consider also that having different bootstrap token makes the UX more complex. 
IMO, considering the short TTL now assigned by default to tokens and all the existing facilities for managing token available under `kubeadm token`, we should keep this simple